### DISCUSSION TITLE
Build navmesh for very big map

### DISCUSSION THREAD

**Discussion by AN3Orik** [recastnavigation/recastnavigation#566] [Q&A]
Hi there.
I trying to use recast/detour for build MMO server side navmesh, but i have a problem: project has ~3000 sectors in .obj format and i can't understand how to create links between sectors/build/store and load this sectors. 
Can someone provide example or roadmap how to do this, please?
Ty in advance.

**Comment by Bromeon** [comment]
You should probably to break it down in a hierarchical way:
* Recast/Detour's nav-meshes support _tiles_, rectangular subdivisions of the world space. They can be created and added one at a time. The (tiled) nav-mesh is essentially a sparse map of tiles, where holes are allowed for subdivisions which don't contain walkable geometry. Thus, you only need to load tiles which are really needed.
* It seems like sectors are another spatial partitioning concepts in your world. Use that to your advantage, e.g. build one nav-mesh per sector and connect the sectors. This connection can be done by laying a graph over the world, with neighboring sectors sharing an edge. To travel, you can first do an A* search on sectors, and only then do a more localized navigation in Detour. Of course it will need special handling of boundaries, shortest-path special cases etc.

It's a complex topic, there are some resources that might help with _hierarchical pathfinding_:
* https://youtu.be/670PcsoSpEY -- general concept
* https://youtu.be/anGdYJu_eH4 -- a concrete game
* https://youtu.be/rg1bLj7bRII -- a concrete, finished game (Castle Story)
* https://factorio.com/blog/post/fff-317 -- devlog for another game (Factorio)
* the rest of the Internet

**Comment by AN3Orik** [comment]
Thanks for answer, but i ask for more detailed example using recast/detour not a theoretical information.
I try to describe my problem more detail.
I has 9 sectors (for example) - each sector is huge geometry object file 
I need build path from point A to point B, how i can with Recast build navmesh for each sector and create links between this sectors to simply use navquery->findPath in my code to find path.

![image](https://user-images.githubusercontent.com/6397450/105223375-bf5beb80-5b6c-11eb-832d-3195afc07bdd.png)

**Comment by Bromeon** [comment]
> I need build path from point A to point B, how i can with Recast build navmesh for each sector and create links between this sectors to simply use navquery->findPath in my code to find path.

That's not possible, you can only use `findPath()` inside a single nav-mesh. Detour's API is not designed for connecting multiple nav-meshes, which is why I suggested approaches to do that on your own.

**Comment by jakobbotsch** [comment]
A tiled navmesh is essentially that. You can build tiles and connect them as a single navmesh. Look at `Sample_TileMesh` in the demo for examples.

**Comment by jakobbotsch** [comment]
As @Bromeon points out, if you want to only load necessary sectors, then it will be more complicated. One possible solution that is not perfect is to simply load the neighboring tiles as you get to the polygons that border that tile in the A* search. This will require some modifications to `findPath`, but nothing major. Alternatively, it might be possible to do this without any modifications using the sliced path finding APIs.
However, this will still not perform as well as a hierarchical solution since the search will explore every polygon required ahead of time.

**Comment by AN3Orik** [comment]
Thanks, i'll try to build tiles for all sector obj files and add to one NavMesh.

**Comment by Bromeon** [comment]
Maybe to get an idea, what size are we talking about? You say each sector contains _huge_ geometry, which gave me the impression that it would be already questionable if a single nav-mesh could fit the entire sector, let alone multiple sectors.

Bear in mind that the pathfinding search speed directly depends on the number of polygons (i.e. A* graph nodes) in your nav-mesh, and may perform extremely bad if the search space is too big. So even if it's technically possible to use a single nav-mesh for the entire world, it won't help you if you cannot find paths in reasonable time.

**Comment by AN3Orik** [comment]
It's a example 4 neighboring sectors. Almost all sectors has just a terrain with some small objects.
[sectors.zip](https://github.com/recastnavigation/recastnavigation/files/5850681/sectors.zip)

**Comment by AN3Orik** [comment]
I tried to create one NavMesh and added to this mesh few generated tiles, but findPath failed.
NavMeshQuery success to find start/end polygons but path interrupted on sectors border. I'm using cellSize 0.2, agentRadius 0.2 and tileSize 640 to guarantee 1 tile per sector (see example obj's files above). 
By the navMesh->addTile() code i see navmesh must be automatically create links between tiles, but i think it's didn't happens. 

Maybe problem with borders at sector boundaries where navmeshes didn't exist and can't be linked with neighbor sector?

**Reply by tlassanske** [reply]
If the NavMesh for each sector has an additional border zone from the adjacent sectors, then the path generated will get as close as it can to the target, causing it to end somewhat inside the other sector, most likely.

You could then immediately do another path search for the adjacent sector from the final point and stitch the two results together. If it is highly likely the destination will be changed well before then, you might defer the second (and third, and fourth) search until the agent reaches each border zone it traverses.

If your world is designed with large sectors that have little to no blockage across sectors, that may be all you need. But if not, you would need a hierarchical approach, as mentioned by prior posters.

One way to do this is to build a very loose NavMesh that covers the entire world. If you use large cell sizes and a large agent radius, the data size can be minimized, though obstacles will be given a wide berth. If the path is entirely within the current sector, you simply search that sector's mesh. If that search fails, or you know the source and destination are in different sectors, then you use the loose mesh to generate the high level path, and then generate a detailed path to the first border, and then another from there to the next border, and so on - likely deferring the detailed searches until the border is reached to defer the expense until it is actually needed.

For example, this image shows 9 sectors with a big mountain range crossing some sector bounds. To path between A and B, the loose mesh is used to generate the thick green path. The mesh offset from the mountain might be as large as a vehicle or a building or whatever needed to keep your build times and data small enough. The thin blue path shows the paths generated for each sector using its own high-resolution NavMesh. Since you are using the sector crossing points from the first search, they will all be valid navigation, but the entire path won't be optimal - but who walks right next to a mountain for miles and miles? Besides, it is likely other obstacles had to be avoided, so the final path might sill look reasonable.
![image](https://github.com/user-attachments/assets/c66623d0-40b6-405b-b96d-18eceffe0dc6)

If you want to spend additional cycles, you could perhaps clean up the high resolution results.
